builtInFunctions/multiESDTNFTTransfer.go

```diff
@@ -0,0 +1,452 @@
+			if err != nil {
+				return nil, err
+			}
+
+			err = e.addNFTToDestination(vmInput.RecipientAddr, acntDst, esdtTransferData, esdtTokenKey, mustVerifyPayable(vmInput, int(minNumOfArguments)))
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			err = addToESDTBalance(acntDst, esdtTokenKey, big.NewInt(0).SetBytes(vmInput.Arguments[tokenStartIndex+2]), e.marshalizer, e.pauseHandler)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		logEntry := newEntryForNFT(vmcommon.BuiltInFunctionMultiESDTNFTTransfer, vmInput.CallerAddr, tokenID, nonce)
+		logEntry.Topics = append(logEntry.Topics, acntDst.AddressBytes())
+		vmOutput.Logs[i] = logEntry
+	}
+
+	// no need to consume gas on destination - sender already paid for it
+	if len(vmInput.Arguments) > int(minNumOfArguments) && vmcommon.IsSmartContractAddress(vmInput.RecipientAddr) {
+		var callArgs [][]byte
+		if len(vmInput.Arguments) > int(minNumOfArguments)+1 {
+			callArgs = vmInput.Arguments[minNumOfArguments+1:]
+		}
+
+		addOutputTransferToVMOutput(
+			vmInput.CallerAddr,
+			string(vmInput.Arguments[minNumOfArguments]),
+			callArgs,
+			vmInput.RecipientAddr,
+			vmInput.GasLocked,
+			vmInput.CallType,
+			vmOutput)
+	}
+
+	return vmOutput, nil
+}
+
+func (e *esdtNFTMultiTransfer) processESDTNFTMultiTransferOnSenderShard(
+	acntSnd vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) (*vmcommon.VMOutput, error) {
+	dstAddress := vmInput.Arguments[0]
+	if len(dstAddress) != len(vmInput.CallerAddr) {
+		return nil, fmt.Errorf("%w, not a valid destination address", ErrInvalidArguments)
+	}
+	if bytes.Equal(dstAddress, vmInput.CallerAddr) {
+		return nil, fmt.Errorf("%w, can not transfer to self", ErrInvalidArguments)
+	}
+	if e.shardCoordinator.ComputeId(dstAddress) == vmcommon.MetachainShardId {
+		return nil, ErrInvalidRcvAddr
+	}
+	numOfTransfers := big.NewInt(0).SetBytes(vmInput.Arguments[1]).Uint64()
+	if numOfTransfers == 0 {
+		return nil, fmt.Errorf("%w, 0 tokens to transfer", ErrInvalidArguments)
+	}
+	minNumOfArguments := numOfTransfers*argumentsPerTransfer + 2
+	if uint64(len(vmInput.Arguments)) < minNumOfArguments {
+		return nil, fmt.Errorf("%w, invalid number of arguments", ErrInvalidArguments)
+	}
+
+	multiTransferCost := numOfTransfers * e.funcGasCost
+	if vmInput.GasProvided < multiTransferCost {
+		return nil, ErrNotEnoughGas
+	}
+
+	verifyPayable := mustVerifyPayable(vmInput, int(minNumOfArguments))
+	acntDst, err := e.loadAccountIfInShard(dstAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	vmOutput := &vmcommon.VMOutput{
+		ReturnCode:   vmcommon.Ok,
+		GasRemaining: vmInput.GasProvided - multiTransferCost,
+		Logs:         make([]*vmcommon.LogEntry, numOfTransfers),
+	}
+
+	startIndex := uint64(2)
+	listEsdtData := make([]*esdt.ESDigitalToken, numOfTransfers)
+	listTokenID := make([][]byte, numOfTransfers)
+	for i := uint64(0); i < numOfTransfers; i++ {
+		tokenStartIndex := startIndex + i*argumentsPerTransfer
```

Comment:
this could have been extracted in a function

Reply:
the underlying execution is extracted to a function. If I extract more there would be a need for a specialized structure which returns everything.